File: Factory/GoFactory/Factory.go
package main

import "fmt"

// Product is the interface for the objects the factory will create
type Product interface {
	Use()
}

// ConcreteProductA is a type of product the factory can create
type ConcreteProductA struct{}

// Use implements the Product interface
func (p *ConcreteProductA) Use() {
	fmt.Println("Using ConcreteProductA")
}

// ConcreteProductB is a type of product the factory can create
type ConcreteProductB struct{}

// Use implements the Product interface
func (p *ConcreteProductB) Use() {
	fmt.Println("Using ConcreteProductB")
}

// Factory is the interface for the object factory
type Factory interface {
	CreateProduct() Product
}

// ConcreteFactoryA is a type of factory that creates ConcreteProductA objects
type ConcreteFactoryA struct{}

// CreateProduct implements the Factory interface
func (f *ConcreteFactoryA) CreateProduct() Product {
	return &ConcreteProductA{}
}

// ConcreteFactoryB is a type of factory that creates ConcreteProductB objects
type ConcreteFactoryB struct{}

// CreateProduct implements the Factory interface
func (f *ConcreteFactoryB) CreateProduct() Product {
	return &ConcreteProductB{}
}

func main() {
	// Create a ConcreteFactoryA and use it to create a ConcreteProductA
	factory := &ConcreteFactoryA{}
	product := factory.CreateProduct()
	product.Use()

	// Create a ConcreteFactoryB and use it to create a ConcreteProductB
	factory = &ConcreteFactoryB{}
	product = factory.CreateProduct()
	product.Use()
}

File: Factory/JavaFactory/Factory.java
import java.util.HashMap;
import java.util.Map;

abstract class Product {
    // Abstract method that must be implemented by concrete products
    public abstract void use();
}

class ConcreteProductA extends Product {
    @Override
    public void use() {
        System.out.println("Using ConcreteProductA");
    }
}

class ConcreteProductB extends Product {
    @Override
    public void use() {
        System.out.println("Using ConcreteProductB");
    }
}

abstract class Creator {
    // Map to store all products
    private static Map<String, Product> products = new HashMap<>();

    // Factory method that returns an instance of a concrete product
    protected abstract Product factoryMethod(String productType);

    // Method to get a product instance by product type
    public Product createProduct(String productType) {
        Product product = products.get(productType);
        if (product == null) {
            product = factoryMethod(productType);
            products.put(productType, product);
        }
        return product;
    }
}

class ConcreteCreatorA extends Creator {
    @Override
    protected Product factoryMethod(String productType) {
        switch (productType) {
            case "A":
                return new ConcreteProductA();
            case "B":
                return new ConcreteProductB();
            default:
                throw new IllegalArgumentException("Invalid product type");
        }
    }
}

class Client {
    public static void main(String[] args) {
        Creator creator = new ConcreteCreatorA();

        Product productA = creator.createProduct("A");
        productA.use();

        Product productB = creator.createProduct("B");
        productB.use();
    }
}
